### PR TITLE
New scheme for non-hash formats, that can exploit same-salt

### PR DESCRIPTION
**Pull request by magnumripper**
See #3497 

WPAPSK and a few other formats have FMT_BLOB at this time.

Notes:
Tunable costs only look at salts, so we can no longer separate WPA1 from WPA2 or 802.11w (OTOH they are not really costs per se - they take about the same time to process). For costs actually "hidden" in a FMT_BLOB we need to look at the blob data (a.k.a binary) and that'd be really hairy, right now I can't see it happening anytime soon. So for now we can't use costs that need to be separated from actual salt. For formats like office, we're storing actual salt, iteration count **and** office version in the "salt" so it's not a a problem there.

### CODE REVIEW

**Comment by magnumripper** [comment]
Format flag added: `FMT_BLOB`.  Apart from that, the format and loader structs are unchanged!  For `FMT_BLOB` we just re-purpose the binary pointer and the `cmp_*` functions.  A new `typedef struct ... fmt_data` is used for blobs, it's always dynamically sized and can be ether tiny_alloc'ed or alloc'ed (cracked alloc'ed blobs will be freed by cracker).  This potentially obsoletes dynamic_salt if/when all such formats are converted to this new style.

File: src/formats.c [hunk 1]
```diff
@@ -1621,7 +1621,7 @@ char *fmt_self_test(struct fmt_main *format, struct db_main *db)
 	self_test_running = 0;
 
 	MEM_FREE(salt_alloc);
-	BLOB_FREE(binary_alloc);
+	//BLOB_FREE(binary_alloc);
```
Comment (magnumripper) [inline src/formats.c]: I can't yet see why this fails with ASan unless commented out. It boils down to this after expansion:
```c
	if (binary_alloc && (format->params.flags & FMT_BLOB) &&
		(((fmt_data*)binary_alloc)->flags & FMT_DATA_ALLOC)) 
			MEM_FREE(((fmt_data*)binary_alloc)->blob);
```

Comment (magnumripper) [reply src/formats.c]: ...Oh and as we all know, `MEM_FREE` boils down to checking that `((fmt_data*)binary_alloc)->blob` is not NULL either. So why the heck does it say BOOM with
```
==49736==The signal is caused by a WRITE memory access.
==49736==Hint: address points to the zero page.
```

**Comment by magnumripper** [comment]
Got another ASan problem during testing. @solardiz what exactly does the following (john proper) code in `crk_process_guess()` do? I don't quite understand it:
```c
        dupe = crk_timestamps[index] == status.crypts;
        crk_timestamps[index] = status.crypts;
```
In order not to miss stuff with `FMT_BLOB` I had to "disable" that, but apparently that was a bit too bold. Since I don't understand the rationale with this I'm not quite sure how I should proceed.

**Review by magnumripper — COMMENTED** [review]
Whatever that `dupe` stuff does, this b2678258f seems to deal with the problem. We'll see how ASan behaves...

**Comment by solardiz** [comment]
> what exactly does the following (john proper) code in `crk_process_guess()` do?

As I recall, it was a (too) clever hack to avoid writing dupes to `john.pot` and to single crack's buffers when the same password is cracked more than once per `crypt_all` (perhaps because the same hash is loaded more than once, which may happen e.g. in single crack mode). However, I cannot confirm it achieving that now when I look at recent code. Maybe we need to test with duplicate hashes and see what happens.

Anyway, I also don't see why this would be relevant to your `FMT_BLOB` work.

> `MEM_FREE` boils down to checking that `((fmt_data*)binary_alloc)->blob` is not NULL

But it wouldn't check whether `(fmt_data*)binary_alloc)` is NULL, and would crash on NULL when trying to access `->blob`.

File: src/wpapsk_fmt_plug.c [hunk 2]
```diff
@@ -212,33 +210,27 @@ struct fmt_main fmt_wpapsk = {
...
 		crypt_all,
 		{
-			get_hash_0,
-			get_hash_1,
-			get_hash_2,
-			get_hash_3,
-			get_hash_4,
-			get_hash_5,
-			get_hash_6
+			fmt_default_get_hash,
```
Comment (solardiz) [inline src/wpapsk_fmt_plug.c]: Looks wrong to me: why use 7 separate `binary_hash_*` functions when we also use `fmt_default_get_hash`, which always returns 0? And, shouldn't there also be hunks removing the now-unused functions `clear_keys` and `get_hash_*`?

**Review by solardiz — COMMENTED** [review]
I don't have time to seriously review this and wasn't planning on including these changes in the very next release, but if you're now in the mood, I won't stop you.

It is indeed good to be able to take advantage of matching ESSIDs. Perhaps we can also take advantage of matching salts with `openssl enc` blobs, in case someone specified a salt manually (unlikely?) and used the same more than once. Maybe there are more examples.

If this interferes with reporting of tunable costs that significantly affect performance, that's really bad, though. For example, what will happen to the `dmg` formats?

**Comment by magnumripper** [comment]
> > what exactly does the following (john proper) code in `crk_process_guess()` do?
> 
> As I recall, it was a (too) clever hack to avoid writing dupes to `john.pot` and to single crack's buffers when the same password is cracked more than once per `crypt_all` (perhaps because the same hash is loaded more than once, which may happen e.g. in single crack mode). However, I cannot confirm it achieving that now when I look at recent code. Maybe we need to test with duplicate hashes and see what happens.

I see. I do think it works as intended...

> Anyway, I also don't see why this would be relevant to your `FMT_BLOB` work.

Because unlike normal hashes' salt and binary, FMT_BLOB can have one salt with multiple different binaries, where some may correspond to one same password (while others may not). Before I excluded FMT_BLOB from this test, some cracks were reported but not written to pot file. BTW my current code seem to disable it at the very right place, with no ill effects.

> > `MEM_FREE` boils down to checking that `((fmt_data*)binary_alloc)->blob` is not NULL
> 
> But it wouldn't check whether `(fmt_data*)binary_alloc)` is NULL, and would crash on NULL when trying to access `->blob`.

Oh, right. I'll re-visit that, thanks. **EDIT** no that was not the problem - the BLOB_FREE macro did check that before calling MEM_FREE. The problem was I did `BLOB_FREE(binary_alloc)` but I should have used `BLOB_FREE(binary_copy)` - the latter is sometimes aligned (to no more than `BINARY_ALIGN`) within the former.

File: src/wpapsk_fmt_plug.c (hunk 2 above)
Comment (magnumripper) [reply src/wpapsk_fmt_plug.c]: With my current hacks (everything is experimental at this point) these are only for **loader** to process pot file entries quicker (it did get WAY quicker with a couple hundred thousand WPA handshake samples). I added code to ignore these functions in cracker when there's no corresponding `get_hash` functions. Not sure if all sizes are required, I just threw them in there.

Yes, there is unused code that is #ifdef'ed out right now (although the `get_hash_*` functions were from shared code). It will all be cleaned up before we're anywhere near merging. I'm mostly interested in any caveats you might see with this non-hash scheme.

**Comment by magnumripper** [comment]
> It is indeed good to be able to take advantage of matching ESSIDs. Perhaps we can also take advantage of matching salts with `openssl enc` blobs, in case someone specified a salt manually (unlikely?) and used the same more than once. Maybe there are more examples.

Not many, but at least a few. I believe Office will/may re-use the salt upon saving an updated document.

> If this interferes with reporting of tunable costs that significantly affect performance, that's really bad, though. For example, what will happen to the `dmg` formats?

As long as a salt-reuse wouldn't be exploitable anyway, we could include `header_ver` & `iter_cnt` in the "salt" but it wouldn't be any better (or worse) than keeping it as is, as far as I can see. We can leave most formats as-is because this new stuff doesn't really clash with the old way of doing it.

**Comment by magnumripper** [comment]
This (the initial commit) is ripe for final reviews now. The rest are format-specific implementations of the new feature.

**Comment by magnumripper** [comment]
I will "have to" merge this once complete or it'll be in the way for future work. If we really don't want it in Jumbo-2 I see no particular problem with branching off a "jumbo" branch with stable/wanted code (as defined by you @solardiz) while I just hack away a bit more freely in bleeding-jumbo (as its name suggests!). You'd get to cherry-pick stuff from bleeding-jumbo to jumbo if/when you want to. How about that?

**Comment by solardiz** [comment]
> I will "have to" merge this once complete

I'm OK with having this merged into the main tree. Just make sure that, in your testing, this doesn't cause breakage of any format, including tunable costs reporting, etc. Thanks!

**Comment by magnumripper** [comment]
Maybe I'll try converting oldOffice/Office as well before merging. Right now I'm not sure any other formats are worth converting.

File: src/wpapsk_fmt_plug.c (hunk 2 above)
Comment (magnumripper) [reply src/wpapsk_fmt_plug.c]: BTW `clear_keys` and `get_hash_*` are (and were) removed, but they were sitting in wpapsk.h

**Comment by magnumripper** [comment]
The office formats could keep all costs features - they're left in the salt struct and this doesn't harm exploiting same-salt.
